planet/src/main/java/com/google/planet/data/ItineraryGenerator.java

```diff
@@ -32,31 +40,59 @@
         return items;
     }
 
     // Function that creates a graph with the events as vertices and the travelling
     // time between the events as edges.
-    // For the prototype, all travelling times are set to 15 minutes 
-    public int[][] getTravelTimeGraph(List<Event> events) {
+    // For the MVP, real time traffic is NOT used.
+    // TODO: handle invalid addresses
+    private int[][] getEventsDistanceMatrix(List<Event> events) {
         int numOfNodes = events.size();
         int[][] travelTimeGraph = new int[numOfNodes][numOfNodes];
-        for (int i = 0; i < travelTimeGraph.length; i++) {
-            for (int j = 0; j < travelTimeGraph[i].length; j++) {
-                if (i == j) {
-                    travelTimeGraph[i][j] = 0;
-                }else{
-                    travelTimeGraph[i][j] = 15;
+
+        String GoogleApiKey = "AIzaSyDK36gDoYgOj4AlbCqh1IuaUuTlcpKF0ns";
+        String[] origins = getListOfAddresses(events);
+        String[] destinations = getListOfAddresses(events);
+        GeoApiContext context = new GeoApiContext.Builder(new GaeRequestHandler.Builder())
+                            .apiKey(GoogleApiKey)
+                            .build();
+       try {
+            DistanceMatrix distanceMatrix = DistanceMatrixApi.getDistanceMatrix(context,
+                                                         origins,
+                                                         destinations).await();
+
+            for (int i = 0; i < distanceMatrix.rows.length; i++) {
+                DistanceMatrixRow currentRow = distanceMatrix.rows[i];
+                for (int j = 0; j < currentRow.elements.length; j++) {
+                    DistanceMatrixElement currentElement = currentRow.elements[j];
+                    Duration duration = currentElement.duration;
+                    travelTimeGraph[i][j] = getTravelDurationInMinutes(duration);
                 }
             }
+            return travelTimeGraph;
+        } catch (Exception e) {
+            System.out.println(e);
+            return travelTimeGraph;
         }
-        return travelTimeGraph;
+    }
+
+    private String[] getListOfAddresses(List<Event> events) {
+        String[] addressList = events.stream()
+                                .map(event -> event.getAddress())
+                                .toArray(String[]::new);
+        return addressList;
+    }
+
+    private int getTravelDurationInMinutes(Duration duration) {
+        int travelDurationInMinutes = (int) Math.round(duration.inSeconds / 60);
+        return travelDurationInMinutes;
     }
 
     // Function that creates an itinerary by scheduling each event in order.
-    public List<ItineraryItem> scheduleItineraryInOrder(List<Event> events, int openningTime, int endingTime, 
+    private List<ItineraryItem> scheduleItineraryInOrder(List<Event> events, int openningTime, int endingTime, 
```

Comment:
it looks like you're implicitly relying on the fact that events and travelTimeGraph are in the same order. That assumption should be explicitly stated.

Even better would be designing a data structure that didn't need that assumption (e.g. by making the graph searchable by name or address.)

Reply:
I added a commit about using the same order!
I think designing a data structure for the travelTimeGraph sounds like a good idea, but currently searching by index is sufficient enough for the algorithm. I can try implementing it next week if it's more necessary then?

Reply:
As it stands, the code is brittle and easy to accidentally break in a future change.

If you're planning to update it in the future, I think it's ok to go ahead and merge as-is.